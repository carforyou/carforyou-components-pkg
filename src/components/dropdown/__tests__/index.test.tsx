import React from "react"
import { cleanup, fireEvent, render, within } from "@testing-library/react"

import Dropdown from "../index"

describe("<Dropdown>", () => {
  const mockedOnSelect = jest.fn()

  const options = [
    { value: 1, name: "One" },
    { value: 2, name: "Two" },
    { value: 3, name: "Three" },
  ]
  const title = "Select number"
  const selected = 1

  const mountWrapper = (data = {}) => {
    const { renderOption } = {
      renderOption: undefined,
      ...data,
    }

    return render(
      <Dropdown
        options={options}
        selected={selected}
        onSelect={mockedOnSelect}
        toggle={({ name }) => (
          <div>
            <span className="font-bold lg:font-regular">{title}</span>:{" "}
            <span className="font-bold">{name}</span>
          </div>
        )}
        renderOption={renderOption}
      />
    )
  }

  const openDropdown = (wrapper) => {
    const { getByText, getByTestId } = wrapper
    const titleNode = getByText(title)

    fireEvent.click(titleNode)
    getByTestId("dropdown-options")
  }

  beforeEach(() => {
    cleanup()
    mockedOnSelect.mockClear()
  })

  describe("correct markup", () => {
    describe("when closed", () => {
      it("renders title", () => {
        const { getByText } = mountWrapper()
        expect(getByText(title)).toMatchSnapshot()
      })

      it("renders selected option", () => {
        const { getByText } = mountWrapper()
        expect(getByText("One")).toMatchSnapshot()
      })
    })

    describe("when opened", () => {
      it("renders title", () => {
        const wrapper = mountWrapper()
        openDropdown(wrapper)

        const { getByText } = wrapper
        expect(getByText(title)).toMatchSnapshot()
      })

      it("renders selected option", () => {
        const wrapper = mountWrapper()
        openDropdown(wrapper)

        const { getByTestId } = wrapper
        const toggle = getByTestId("dropdown-toggle")
        expect(within(toggle).getByText("One")).toMatchSnapshot()
      })

      it("renders all options", () => {
        const wrapper = mountWrapper()
        openDropdown(wrapper)

        const { getByTestId } = mountWrapper()
        const list = getByTestId("dropdown-options")
        options.forEach(({ name }) => {
          expect(within(list).getByText(name)).toMatchSnapshot()
        })
      })
    })
  })

  it("passes selection to callback", () => {
    const wrapper = mountWrapper()
    openDropdown(wrapper)

    const { getByText } = wrapper
    const optionNode = getByText("Two")
    fireEvent.click(optionNode)

    expect(mockedOnSelect).toBeCalledWith(2)
  })

  it("closes after selecting option", async () => {
    const wrapper = mountWrapper()
    openDropdown(wrapper)

    const { getByText, queryAllByTestId } = wrapper
    const optionNode = getByText("Two")
    fireEvent.click(optionNode)

    expect(queryAllByTestId("dropdown-options")).toEqual([])
  })

  it("accepts custom render for options", () => {
    const wrapper = mountWrapper({
      renderOption: ({ name, value }) => (
        <div>
          {name} ({value})
        </div>
      ),
    })

    openDropdown(wrapper)
    const { getByText } = wrapper

    options.forEach(({ value, name }) =>
      expect(getByText(`${name} (${value})`))
    )
  })
})
