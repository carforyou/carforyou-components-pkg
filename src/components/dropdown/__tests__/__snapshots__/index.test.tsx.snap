// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown> correct markup when closed renders selected option 1`] = `
<span
  class="font-bold"
>
  One
</span>
`;

exports[`<Dropdown> correct markup when closed renders title 1`] = `
<span
  class="font-bold lg:font-regular"
>
  Select number
</span>
`;

exports[`<Dropdown> correct markup when opened renders all options 1`] = `
<li
  aria-selected="false"
  class="hover:bg-grey-bright transition-2 px-20 py-10 cursor-pointer font-bold text-teal"
  data-testid="One"
  id="downshift-4-item-0"
  role="option"
>
  One
</li>
`;

exports[`<Dropdown> correct markup when opened renders all options 2`] = `
<li
  aria-selected="false"
  class="hover:bg-grey-bright transition-2 px-20 py-10 cursor-pointer"
  data-testid="Two"
  id="downshift-4-item-1"
  role="option"
>
  Two
</li>
`;

exports[`<Dropdown> correct markup when opened renders all options 3`] = `
<li
  aria-selected="false"
  class="hover:bg-grey-bright transition-2 px-20 py-10 cursor-pointer"
  data-testid="Three"
  id="downshift-4-item-2"
  role="option"
>
  Three
</li>
`;

exports[`<Dropdown> correct markup when opened renders selected option 1`] = `
<span
  class="font-bold"
>
  One
</span>
`;

exports[`<Dropdown> correct markup when opened renders title 1`] = `
<span
  class="font-bold lg:font-regular"
>
  Select number
</span>
`;
